hive4-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java

```diff
@@ -0,0 +1,252 @@
+      this.executorService.submit(() -> server.serve());
+
+      // in Hive3, setting this as a system prop ensures that it will be picked up whenever a new
+      // HiveConf is created
+      System.setProperty(
+          "hive.metastore.uris", hiveConf.getVar(HiveConf.getConfVars("hive.metastore.uris")));
+
+      this.clientPool = new HiveClientPool(1, hiveConf);
+    } catch (Exception e) {
+      throw new RuntimeException("Cannot start TestHiveMetastore", e);
+    }
+  }
+
+  public void stop() throws Exception {
+    reset();
+    if (clientPool != null) {
+      clientPool.close();
+    }
+    if (server != null) {
+      server.stop();
+    }
+    if (executorService != null) {
+      executorService.shutdown();
+    }
+    if (baseHandler != null) {
+      baseHandler.shutdown();
+    }
+    ThreadPool.shutdown();
+  }
+
+  public HiveConf hiveConf() {
+    return hiveConf;
+  }
+
+  public String getDatabasePath(String dbName) {
+    File dbDir = new File(HIVE_LOCAL_DIR, dbName + ".db");
+    return dbDir.getPath();
+  }
+
+  public void reset() throws Exception {
+    if (clientPool != null) {
+      for (String dbName : clientPool.run(client -> client.getAllDatabases())) {
+        for (String tblName : clientPool.run(client -> client.getAllTables(dbName))) {
+          clientPool.run(
+              client -> {
+                client.dropTable(dbName, tblName, true, true, true);
+                return null;
+              });
+        }
+
+        if (!DEFAULT_DATABASE_NAME.equals(dbName)) {
+          // Drop cascade, functions dropped by cascade
+          clientPool.run(
+              client -> {
+                client.dropDatabase(dbName, true, true, true);
+                return null;
+              });
+        }
+      }
+    }
+
+    Path warehouseRoot = new Path(HIVE_LOCAL_DIR.getAbsolutePath());
+    FileSystem fs = Util.getFs(warehouseRoot, hiveConf);
+    for (FileStatus fileStatus : fs.listStatus(warehouseRoot)) {
+      if (!fileStatus.getPath().getName().equals("derby.log")
+          && !fileStatus.getPath().getName().equals("metastore_db")) {
+        fs.delete(fileStatus.getPath(), true);
+      }
+    }
+  }
+
+  public Table getTable(String dbName, String tableName) throws TException, InterruptedException {
+    return clientPool.run(client -> client.getTable(dbName, tableName));
+  }
+
+  public Table getTable(TableIdentifier identifier) throws TException, InterruptedException {
+    return getTable(identifier.namespace().toString(), identifier.name());
+  }
+
+  private TServer newThriftServer(TServerSocket socket, int poolSize, HiveConf conf)
+      throws Exception {
+    HiveConf serverConf = new HiveConf(conf);
+    serverConf.set("javax.jdo.option.ConnectionURL", "jdbc:derby:" + DERBY_PATH + ";create=true");
+    baseHandler = new HMSHandler("new db based metaserver", serverConf);
+    IHMSHandler handler = HMSHandlerProxyFactory.getProxy(serverConf, baseHandler, false);
```

Comment:
Since this file is only for Hive 4, we don't need to use the `DynConstructors` and `DynMethods` machinery. We simply call the relevant constructor and static method directly!